rust/feature-flags/src/flag_matching.rs

```diff
@@ -108,39 +130,70 @@ impl FeatureFlagMatcher {
         condition: &FlagGroupType,
         _index: usize,
-    ) -> (bool, String) {
+    ) -> Result<(bool, String), FlagError> {
         let rollout_percentage = condition.rollout_percentage.unwrap_or(100.0);
-        let mut condition_match = true;
-
-        if let Some(ref properties) = condition.properties {
+        if let Some(properties) = &condition.properties {
             if properties.is_empty() {
-                condition_match = true;
-            } else {
-                // TODO: First handle given override properties before going to db
-                let target_properties = self
-                    .get_person_properties(feature_flag.team_id, self.distinct_id.clone())
-                    .await
-                    .unwrap_or_default();
-                // TODO: Handle db issues / person not found
-
-                condition_match = properties.iter().all(|property| {
-                    match_property(property, &target_properties, false).unwrap_or(false)
-                });
+                return Ok(self.check_rollout(feature_flag, rollout_percentage));
             }
-        };
 
-        if !condition_match {
-            return (false, "NO_CONDITION_MATCH".to_string());
-        } else if rollout_percentage == 100.0 {
-            // TODO: Check floating point schenanigans if any
-            return (true, "CONDITION_MATCH".to_string());
+            let target_properties = self.get_target_properties(feature_flag, properties).await?;
+
+            if !self.all_properties_match(properties, &target_properties) {
+                return Ok((false, "NO_CONDITION_MATCH".to_string()));
+            }
         }
 
-        if self.get_hash(feature_flag, "") > (rollout_percentage / 100.0) {
-            return (false, "OUT_OF_ROLLOUT_BOUND".to_string());
+        Ok(self.check_rollout(feature_flag, rollout_percentage))
+    }
+
+    async fn get_target_properties(
+        &mut self,
+        feature_flag: &FeatureFlag,
+        properties: &[PropertyFilter],
+    ) -> Result<HashMap<String, Value>, FlagError> {
+        self.get_person_properties(feature_flag.team_id, properties)
+            .await
+        // TODO handle group properties, will go something like this
+        // if let Some(group_index) = feature_flag.get_group_type_index() {
+        //     self.get_group_properties(feature_flag.team_id, group_index, properties)
+        // } else {
+        //     self.get_person_properties(feature_flag.team_id, properties)
+        //         .await
+        // }
+    }
+
+    async fn get_person_properties(
+        &mut self,
+        team_id: i32,
+        properties: &[PropertyFilter],
+    ) -> Result<HashMap<String, Value>, FlagError> {
+        if let Some(person_overrides) = &self.person_property_overrides {
+            // Check if all required properties are present in the overrides
+            // and none of them are of type "cohort"
+            let should_prefer_overrides = properties
```

Comment:
metrics are coming in a later PR, but I'll add a TODO to add a counter here, it's a good place for it.

Reply:
> Interesting, to confirm, this checks that all the person properties from the request payload match the spec for a feature flag and short-circuits the database call?

Yes, that's how the functionality works.

Reply:
We automatically do this for any call from the SDKs where props have been added. It's also the only way to subvert the delay with ingesting properties problem.

Reply:
also note that since conditions are OR'ed, as long as even one property is sent where the condition matches, you can short circuit flag evaluation without needing all properties.